cli: validate commands

lread opened the pull request:
Bumped babashka.cli to get fix for table formatting.

Squint has a couple of cli oddities:
- `compile` command does not have to be specified for compile
- `-e` acts like a command but looks like an option

I added upfront processing so these oddities become normal commands for babashka.cli.

Changes:
- Any unrecognized option is a validation error
- Any missing required option is a validation error
- An invalid number of args is a validation error
- On validation error squint exits with code 1
- `nrepl-server` is now described/exposed
- `compile` `:copy-resources` option is now described/exposed
- `-e` `--repl` option is now described/exposed
- `run` now exposes/describes options, same as `compile` but without `--paths` and `copy-resources`
- Usage for `watch` is same as `compile` except `:paths` is required
- Added a few cli sanity tests

Closes #691

Please answer the following questions and leave the below in as part of your PR.

- [x] This PR corresponds to an [issue with a clear problem statement](https://github.com/babashka/babashka/blob/master/doc/dev.md#start-with-an-issue-before-writing-code).

- [x] This PR contains a [test](https://github.com/babashka/babashka/blob/master/doc/dev.md#tests) to prevent against future regressions

- [x] I have updated the [CHANGELOG.md](https://github.com/squint-cljs/squint/blob/master/CHANGELOG.md) file with a description of the addressed issue.

lread committed:
cli: validate commands

Bumped babashka.cli to get fix for table formatting.

Squint has a couple of cli oddities:
- `compile` command does not have to be specified for compile
- `-e` acts like a command but looks like an option

I added upfront processing so these oddities become normal commands for
babashka.cli.

Changes:
- Any unrecognized option is a validation error
- Any missing required option is a validation error
- An invalid number of args is a validation error
- On validation error squint exits with code 1
- `nrepl-server` is now described/exposed
- `compile` `:copy-resources` option is now described/exposed
- `-e` `--repl` option is now described/exposed
- `run` now exposes/describes options, same as `compile` but without
`--paths` and `copy-resources`
- Usage for `watch` is same as `compile` except `:paths` is required
- Added a few cli sanity tests

Closes #691

lread committed:
docs: update changelog

lread committed:
Merge remote-tracking branch 'upstream/main' into lread/validate-cli

* upstream/main:
  Fix #697 (#699)
  Fix typo in resource copy log line (#698)

borkdude commented:
One thought about the ANSI output. When you pipe the output to a text file you see this:

[31mERROR[0m: Unrecognized option: --xrepl

Is that a problem? ANSI output only makes sense on the terminal.

I have an open issue about detecting "terminal" here:

https://github.com/babashka/babashka/issues/1299

In jet I detect terminal in a reliable way here:

https://github.com/borkdude/jet/blob/master/src-java/org/babashka/CLibrary.java

but it can also be done using `(System/console)`. The only downside to `(System/console)` is that when you pipe stdin from a non-terminal, it returns `nil`:

```
bb -e '(System/console)' <<< 1
```

but perhaps that's just good enough.

lread commented:
Good point on ANSI escapes. Node itself outputs color. Let's see what it does for comparison:
^
```
node lib/cli.js compile missing.cljs
```
<img width="1260" height="504" alt="image" src="https://github.com/user-attachments/assets/7dc39e36-2e77-4942-b2ea-a49fa6a382da" />

This is written to stderr, if we redirect:
```
node lib/cli.js compile missing.cljs &> foo
```
The ansi codes are not there.

Does it respect [NO_COLOR](https://no-color.org/)?
```
NO_COLOR=true node lib/cli.js compile missing.cljs
```
It does. Maybe supporting `NO_COLOR` env var is good enough?

lread committed:
review feedback: ANSI colour codes

Mimic node. Don't use ANSI colour codes if any true:
1) `NO_COLOR` env var is set
2) not in an interactive terminal (ex. redirecting output)

lread committed:
review feedback: allow ANSI color codes on Windows

I remembered incorrectly, ANSI colour codes do work on Cmd and
Powershell on Windows.

lread committed:
review feedback: paths has a default of: ., src

This was only expressed internally, but needs to be also expressed in
the cli so that:
- the user can learn what the default is
- the cli uses paths itself for compile, so those defaults need to be
applied earlier for it.

This means squint no longer has any required opts.
Supporting code turfed.

This also means we no longer have our failure check for both `<files>`
and `--paths` being specified for `compile`. Updated usage note to
explain how `<files>` overrides `--paths` and `:paths`.

This allowed me to turf post-validate supporting code.

borkdude committed:
cli: drop deferred-default workaround, use :exec-args directly

babashka.cli's :exec-args already implements the precedence layer
needed for squint.edn: CLI > :exec-args > spec :default. spec->opts
merges spec defaults into :exec-args with user-supplied :exec-args
winning, so passing squint.edn as :exec-args gives the right result
without any deferral/re-application gymnastics.

Removes:
- spec-opt-key (defer/enable :default rename dance)
- apply-opt-defaults (manual default application)
- clojure.set require (only used by spec-opt-key)

Net: -23 lines in cli.cljs, same observable behavior. All node_cli_tests
still pass.

borkdude committed:
cli: select-keys squint.edn to spec keys before :exec-args

So stray keys in squint.edn don't trip :restrict validation,
matching prior PR behavior.

borkdude committed:
ci: bump action versions (fix Windows Bad credentials)

- DeLaGuardo/setup-clojure 12.5 -> 13.4
- actions/checkout@v2 -> v4
- actions/setup-java@v2 -> v4
- actions/setup-node@v3 -> v4

setup-clojure 12.5 hits Bad credentials on Windows runners. Also
clears Node.js 20 deprecation warnings.

borkdude committed:
ci: cli: latest (1.10.3.1040 not in casselc/clj-msi)

borkdude committed:
ci: setup-clojure github-token: '' (skip auth; fixes Bad credentials)

Per DeLaGuardo/setup-clojure#87: passing an empty github-token skips
the Authorization header for GH API calls, which avoids the recent
'Bad credentials' failures (token rejected). Public artifacts (Clojure
CLI, bb) don't need auth.

borkdude committed:
ci: setup-clojure@13.4 + cli: latest

borkdude committed:
ci: pin cli + bb versions to skip GH API calls

borkdude committed:
ci: explicit GITHUB_TOKEN + permissions block for setup-clojure

borkdude committed:
ci: remove permissions block (restores default token scopes)

borkdude committed:
Merge remote-tracking branch 'origin/main' into borkdude/use-exec-args

# Conflicts:
#	.github/workflows/ci.yml
#	CHANGELOG.md
#	deps.edn
#	src/squint/internal/cli.cljs

borkdude committed:
test: bump expected compile counts after main merge

test-project/src gained macros_impl.cljc and macros_transitive.cljc;
adjust the squint-edn-* tests' source counts by +2.

borkdude committed:
cli: use :implicit-true flag instead of (= "true" value)

bb cli expose branch surfaces :implicit-true in coerce error data so
callers can detect 'user typed --foo with no value' without pattern
matching on the parser-internal stringified "true". Bumps cli to the
expose branch sha.

borkdude committed:
cli: bump babashka.cli to 0.9.68, use parse-opts* for peeks

- Switch deps.edn to released 0.9.68 (was pinned to expose branch sha).
- Add peek-cli-args using cli/parse-opts*. Peek call sites
  (support-implicit-compile, treat-eval-opt-as-command, *-help-requested,
  cmd-def-from-cli-args) now use it. parse-args' auto-coerce would
  turn '-e false' into ':e false' (boolean), making the
  presence check (:e opts) falsy. parse-opts* keeps values as strings.

borkdude committed:
cli: fix --help detection after parse-opts* swap

parse-opts* keeps option values as raw strings, so :help is "true"
not true. The (= {:help true} opts) check broke. Replace with a
shape check that doesn't depend on the value.

borkdude commented:
Hey @lread - I merged a lot of your commits on main (squashed) and put my own on top while I was standardizing the new bb.cli help output for subcommands. I think this covers most of what you had in mind.
